### PR TITLE
fix: allow building eigen for oss contributors

### PR DESCRIPTION
**Pull request by brainbicycle**
This PR resolves [MOPLAT-6]

### Description

Resolves a few issues preventing building and running for OSS contributors.

- Adds an OSS flag
- Adds instructions for getting a Mapbox token
- Don't initialize firebase app for open source contributors
- Don't initialize unleash for open source contributors
- Don't initialize sentry for open source contributors

Still to do: 
- [x] Add back the google files + add some phony files to allow compilation for OSS
- [x] Add the old open source client id and secret to .env.example
- [x] Clean up some of the handling, don't love the unleash handling and scattered if checks
- [ ] Handle failing relay requests once logged in (friendly error message?)
- [ ] Document known issues with running this way 

### PR Checklist

- [ ] I tested my changes on **iOS** / **Android**.
- [x] I added screenshots or videos to illustrate my changes.
- [x] I added Tests and Stories for my changes.
- [x] I added an [app state migration].
- [x] I hid my changes behind a [feature flag].

### To the reviewers 👀

- [ ] I would like **at least one** of the reviewers to run this PR on the simulator or device.

<details><summary>Changelog updates</summary>

#### Cross-platform user-facing changes

-

#### iOS user-facing changes

-

#### Android user-facing changes

-

#### Dev changes

- Allow compiling and running app for open source contributors - Brian 

</details>

Need help with something? Have a look at our [docs], or get in touch with us.

[app state migration]: ../blob/main/docs/adding_state_migrations.md
[feature flag]: ../blob/main/docs/developing_a_feature.md
[docs]: ../blob/main/docs/README.md

[MOPLAT-6]: https://artsyproduct.atlassian.net/browse/MOPLAT-6?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: ios/Artsy/ArtsyUIFonts/UIFont+ArtsyFonts.m [hunk 1]
```diff
@@ -39,7 +35,10 @@ + (void)loadFontWithFileName:(NSString *)fontFileName extension:(NSString *)exte
         }
 
         CFRelease(font);
+    } else {
+        @throw [NSException exceptionWithName:NSInternalInconsistencyException reason:@"Font failed to load if you are an open source contributor set OSS=true in your ENV file." userInfo:@{}];
```
Comment (olerichter00) [inline ios/Artsy/ArtsyUIFonts/UIFont+ArtsyFonts.m:39]: Does `env=true` work as well or should we adjust the error message? 🤔 


```suggestion
        @throw [NSException exceptionWithName:NSInternalInconsistencyException reason:@"Font failed to load if you are an open source contributor set OSS=True in your ENV file." userInfo:@{}];
```

File: src/app/system/errorReporting/sentrySetup.ts [hunk 2]
```diff
@@ -13,19 +13,28 @@ export const eigenSentryReleaseName = () => {
...
+  }
+
+  if (!sentryDSN) {
     console.error("Sentry DSN not set!!")
     return
   }
 
   Sentry.init({
-    dsn: Config.SENTRY_DSN,
+    dsn: !ossUser ? sentryDSN : undefined,
```
Comment (brainbicycle) [inline src/app/system/errorReporting/sentrySetup.ts:37]: can just pass the dsn here we will return before hitting if they are an oss user

**Comment by gkartalis** [comment]
just adding [this issue](https://github.com/artsy/eigen/issues/8411) that was opened yesterday cause its related

File: ios/Artsy/ArtsyUIFonts/UIFont+ArtsyFonts.m [hunk 3]
```diff
@@ -1,16 +1,12 @@
 // NOTE: This file should be a copy and paste of
 //       https://github.com/artsy/Artsy-OSSUIFonts/blob/master/Pod/Classes/UIFont+ArtsyFonts.m
-//       but with useClosedFonts = YES;
 
 #import "UIFont+ArtsyFonts.h"
 #import <CoreText/CoreText.h>
+#import <react-native-config/ReactNativeConfig.h>
```
Comment (brainbicycle) [inline ios/Artsy/ArtsyUIFonts/UIFont+ArtsyFonts.m:6]: don't need this
```suggestion
```

File: scripts/setup-oss-fonts [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+mkdir -p ./fonts
+
+for i in "${fonts[@]}"
+do
+   aws s3 cp "s3://artsy-webfonts-production/$i.otf" "./fonts/$i.otf"
+   cp "./fonts/$i.otf" "$ANDROID_FONT_DIR"
+   cp "./fonts/$i.otf" "$IOS_FONT_DIR"
+done
+
+rm -rf ./fonts
```
Comment (damassi) [inline scripts/setup-oss-fonts:18]: 👍

File: src/app/App.tsx [hunk 5]
```diff
@@ -80,6 +81,9 @@ const Main = () => {
   useRageShakeDevMenu()
   useDebugging()
   useEffect(() => {
+    if (Config.OSS === "True") {
```
Comment (damassi) [inline src/app/App.tsx:84]: maybe just `true`? That's more common from env files

File: src/app/utils/experiments/unleashClient.ts [hunk 6]
```diff
@@ -12,6 +12,10 @@ import { UnleashClient } from "unleash-proxy-client"
  * or create a new one in the requested env.
  */
 export function getUnleashClient(env?: "production" | "staging", userId?: string | null) {
+  if (Config.OSS === "True") {
+    return null
```
Comment (damassi) [inline src/app/utils/experiments/unleashClient.ts:16]: I wonder how we can avoid all of the extra `?` that are now required in the code. Our unleash usage is going to increase  pretty dramatically through time, especially once we retire echo. 

Non-blocking, but worth thinking about! Maybe we can just turn it off via the api? Like, initialize the client, but then immediately call [stop](https://github.com/Unleash/unleash-proxy-client-js/blob/main/src/index.ts#L317)? 

Or catch errors with logbox? Not too sure.

Comment (brainbicycle) [reply src/app/utils/experiments/unleashClient.ts:16]: the problem is if you try to initialize the client without keys or with bad keys it crashes I believe. I can try again but at least the unleash stuff is abstracted behind the hook so really that is the only place the ? should affect

Comment (damassi) [reply src/app/utils/experiments/unleashClient.ts:16]: Yah its not so bad, just wondering if there might be some easy thing we can do like maybe return a stub api and cast it with an unleast type:

```tsx
if (oss) {
  return {
    foo: x => x,
    bar: x => x, 
  } as unknown as UnleashClient 
}
```

 But that's probably no prettier. In any case, whatever y'all think is best 👍

Comment (brainbicycle) [reply src/app/utils/experiments/unleashClient.ts:16]: oh that is an interesting thought, I can try

File: src/app/utils/experiments/unleashClient.ts [hunk 7]
```diff
@@ -4,14 +4,31 @@ import { nullToUndef } from "app/utils/nullAndUndef"
...
 
+// We want to return a phony unleash client for oss builds
+// Some typescript magic so we only have to specify public fields
+type PublicFields<T> = {
+  [K in keyof T]: T[K] extends { [P in keyof T[K]]: T[K][P] } ? K : never
+}[keyof T]
+
+type PublicType<T> = Pick<T, PublicFields<T>>
+
+export type PublicUnleashClient = PublicType<UnleashClient>
```
Comment (brainbicycle) [inline src/app/utils/experiments/unleashClient.ts:15]: chatGPT helped me write this one :)

File: src/app/utils/experiments/unleashClient.ts (hunk 6 above)
Comment (brainbicycle) [reply src/app/utils/experiments/unleashClient.ts]: this seems like it will work, take a look and let me know what you think 🙏

**Comment by MounirDhahri** [comment]
👏👏👏

**Comment by SamRozen** [comment]
🔥 Nice one! Thanks @brainbicycle for taking this on.

**Pull request merged**